tools/outline_proxy_controller/outline_controller_server.cpp

```diff
@@ -95,29 +109,67 @@ boost::asio::awaitable<void> OutlineClientSession::Start() {
 
       logger.debug("handling client request \"" + client_command + "\"...");
       auto result = RunClientCommand(request_obj);
-
-      // TODO: replace the following code with a json library to handle special characters
-      response << "{\"statusCode\": " << result.status
-               << ",\"returnValue\": \"" << result.result << "\""
-               << ",\"action\": \"" << result.action << "\"}";
-      co_await async_write(channel_, buffer(response.str()), use_awaitable);
-      logger.debug("Wrote back \"" + response.str() + "\" to unix socket");
+      co_await SendResponse(result);
 
       client_command.clear();
-      response.str(std::string{});
     }
   } catch (const std::exception& e) {
     channel_.close();
   }
 }
 
+boost::asio::awaitable<void> OutlineClientSession::MonitorNetworkChanges() {
+  using namespace boost::asio;
+
+  auto executor = co_await this_coro::executor;
+  try {
+    NetworkMonitor net_monitor{executor};
+    for (;;) {
+      // No need to detect the race condition of whether the change is caused by
+      // us (which might cause infinite loop) because we are running in a single-thread
+      // coroutine (see OutlineClientSession::Start()), the code after co_await will be
+      // executed either before RunClientCommand() or after RunClientCommand().
+      auto evt = co_await net_monitor.WaitForChangeEvent();
+      if (evt.HasRoutingChanged()) {
+        if (outline_controller_->IsOutlineRoutingPolluted()) {
```

Comment:
I guess this is also a nice guard against multiple changes detected at the same time.

Reply:
Yes it is, but it did not help a lot. We are still receiving more than 6 routing changes for one single "CONNECT" operation.